File: api/plugins/bingo/dotnet/BingoObject.cs
using System;
using System.Collections;

namespace com.ggasoftware.indigo
{
    /// <summary>
    /// Bingo search object
    /// </summary>
    public unsafe class BingoObject : IDisposable
    {
        private int _id;
        private Indigo _indigo;
        private BingoLib _bingoLib;
        private IDisposable _reference;

        internal BingoObject(int id, Indigo indigo, BingoLib bingo_lib)
        {
            this._id = id;
            this._indigo = indigo;
            this._bingoLib = bingo_lib;
        }

        /// <summary>
        /// Destructor
        /// </summary>
        ~BingoObject()
        {
            Dispose();
        }
        /// <summary>
        /// </summary>
        public void Dispose()
        {
            if (_id >= 0)
            {
                Bingo.checkResult(_indigo, _bingoLib.bingoEndSearch(_id));
                _id = -1;
            }
        }

        /// <summary>
        /// Close method
        /// </summary>
        public void close()
        {
            Dispose();
        }

        internal int id
        {
            get { return _id; }
            set { _id = value; }
        }

        /// <summary>
        /// Method to move to the next record
        /// </summary>
        /// <returns>True if there are any more records</returns>
        public bool next()
        {
            return (Bingo.checkResult(_indigo, _bingoLib.bingoNext(_id)) == 1) ? true : false;
        }

        /// <summary>
        /// Method to return current record id. Should be called after next() method.
        /// </summary>
        /// <returns>Record id</returns>
        public int getCurrentId()
        {
            return Bingo.checkResult(_indigo, _bingoLib.bingoGetCurrentId(_id));
        }

        /// <summary>
        /// Returns a shared IndigoObject for the matched target
        /// </summary>
        /// <returns>Shared Indigo object for the current search operation</returns>
        public IndigoObject getIndigoObject()
        {
            IndigoObject res = new IndigoObject(_indigo, Bingo.checkResult(_indigo, _bingoLib.bingoGetObject(_id)));
            _reference = res;
            return res;
        }
        
        /// <summary>
        /// Method to estimate remaining hits count
        /// </summary>
        /// <returns>Estimated hits count</returns>
        public int estimateRemainingResultsCount ()
        {
            return Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingResultsCount(_id));
        }
        
        /// <summary>
        /// Method to estimate remaining hits count error
        /// </summary>
        /// <returns>Estimated hits count error</returns>
        public int estimateRemainingResultsCountError ()
        {
            return Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingResultsCountError(_id));
        }
        
        /// <summary>
        /// Method to estimate remaining search time
        /// </summary>
        /// <returns>Estimated remainin search time</returns>
        public float estimateRemainingTime ()
        {
            float esimated_time;
            Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingTime(_id, &esimated_time));
            return esimated_time;
        }
    }
}

File: api/plugins/bingo/java/src/main/java/com/ggasoftware/indigo/BingoObject.java
package com.ggasoftware.indigo;

import com.sun.jna.ptr.FloatByReference;

/**
	Bingo search object
*/
public class BingoObject {
	private int _id;
	private Indigo _indigo;
	private BingoLib _bingoLib;
    private Object _reference;

	BingoObject(int id, Indigo indigo, BingoLib bingo_lib) {
		_id = id;
		_indigo = indigo;
		_bingoLib = bingo_lib;
	}

	protected void finalize() {
		dispose();
	}

	public void dispose() {
        if (_id >= 0) {
	       Bingo.checkResult(_indigo, _bingoLib.bingoEndSearch(_id));
            _id = -1;
        }
	}

	public void close() {
		dispose();
	}


	/**
        Move to the next record

        return True if there are any more records
    */
    public boolean next() {
        _indigo.setSessionID();
        return (_bingoLib.bingoNext(_id) == 1);
	}

	/**
	   Return current record id. Should be called after next() method.

	   @return Record id
    */
	public int getCurrentId() {
        _indigo.setSessionID();
        return Bingo.checkResult(_indigo, _bingoLib.bingoGetCurrentId(_id));
	}

    /**
        Return current similarity value. Should be called after next() method.

        @return Similarity value
    */
	public float getCurrentSimilarityValue()
	{
        _indigo.setSessionID();
        return Bingo.checkResult(_indigo, _bingoLib.bingoGetCurrentSimilarityValue(_id));
	}

    /**
	   Return a shared IndigoObject for the matched target

	   @return Shared Indigo object for the current search operation
    **/
	public IndigoObject getIndigoObject() {
        _indigo.setSessionID();
        IndigoObject res = new IndigoObject(_indigo, Bingo.checkResult(_indigo, _bingoLib.bingoGetObject(_id)));
        _reference = res;
        return res;
	}

	/**
	   Estimate remaining hits count

	   @return Estimated hits count
    */
	public int estimateRemainingResultsCount() {
        _indigo.setSessionID();
        return Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingResultsCount(_id));
	}

	/**
        Estimate remaining hits count error

	   @return Estimated hits count error
    */
	public int estimateRemainingResultsCountError() {
        _indigo.setSessionID();
        return Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingResultsCountError(_id));
	}

    /**
        Estimate remaining search time

       @return Estimated remaining search time
    */
	public float estimateRemainingTime () {
        FloatByReference estimated_time = new FloatByReference();
        _indigo.setSessionID();
        Bingo.checkResult(_indigo, _bingoLib.bingoEstimateRemainingTime(_id, estimated_time));
        return estimated_time.getValue();
	}
}
